compilers/concrete-compiler/compiler/include/concretelang/Transforms/TypeInferenceRewriter.h

```diff
@@ -0,0 +1,327 @@
+    if (!llvm::any_of(op->getOperands(),
+                      [&](mlir::Value v) {
+                        return typeResolver.isUnresolvedType(v.getType());
+                      }) &&
+        !llvm::any_of(op->getResults(), [&](mlir::Value v) {
+          return typeResolver.isUnresolvedType(v.getType());
+        })) {
+      resolvedOperandTypes = llvm::to_vector(op->getOperandTypes());
+      resolvedResultTypes = llvm::to_vector(op->getResultTypes());
+    } else if (op->hasTrait<mlir::OpTrait::ReturnLike>()) {
+      // Return-like ops are a bit special, since they simply forward
+      // values upwards. The types inferred for the operands may come
+      // from producers, which are in the same block, while the result
+      // types of the parent op may have been deduced from producers
+      // or consumers of the block containing the parent operation.
+      //
+      // Blindly taking the operand types may thus result in a
+      // mismatch between the final return types of the parent op and
+      // the operand types of the return-like op.
+      //
+      // In the general case, simply take the result types of the
+      // parent operation, which, at this point, has already been
+      // partially rewritten before recursing into this rewrite call.
+      //
+      // Functions are a bit diffent though, since the types of the
+      // results are contained in a function type and not in the
+      // result types.
+      mlir::Operation *newParent = mapping.lookup(op->getParentOp());
+
+      if (llvm::isa<mlir::func::ReturnOp>(op)) {
+        mlir::func::FuncOp newParentFunc =
+            llvm::dyn_cast<mlir::func::FuncOp>(newParent);
+        resolvedOperandTypes =
+            llvm::to_vector(newParentFunc.getFunctionType().getResults());
+      } else {
+        // Look up new parent op and use the return types, since these
+        // are the authorative types obtained from the last invocation
+        // of the type resolver
+        resolvedOperandTypes = llvm::to_vector(newParent->getResultTypes());
+      }
+    } else {
+      LocalInferenceState inferredTypes =
+          TypeInferenceUtils::getLocalInferenceState(solver, op);
+
+      typeResolver.resolve(op, resolvedTypes, inferredTypes);
+
+      resolvedOperandTypes = resolvedTypes.lookup(op->getOperands());
+      resolvedResultTypes = resolvedTypes.lookup(op->getResults());
+
+      if (checkAllResolved(op, resolvedOperandTypes, "operand").failed() ||
+          checkAllResolved(op, resolvedResultTypes, "return value").failed()) {
+        return mlir::failure();
+      }
+    }
+
+    // The types retrieved through the last invocation of the type
+    // resolver do not necessarily match the types of the operands
+    // produced by operations that have already been rewritten, since
+    // the resolved types for the producers may originate from
+    // iterations of the type inference analysis of different
+    // operations.
```

Comment:
I definetely don't get why the type resolver must be called one last time on the function haha.

Reply:
When invoking the type resolver on an operation, the resolver yields a certain set of types based on the local rules for the operation. However, this is only coherent locally for the operation. The locally inferred type for an operand may be different from the type that was inferred for the result of a producing operation. Therefore, the types from the global inference state cannot be applied "blindly" to the operations.